cmd/csi-node-driver-registrar/main.go

```diff
@@ -51,7 +52,13 @@ var (
 	healthzPort             = flag.Int("health-port", 0, "(deprecated) TCP port for healthz requests. Set to 0 to disable the healthz server. Only one of `--health-port` and `--http-endpoint` can be set.")
 	httpEndpoint            = flag.String("http-endpoint", "", "The TCP network address where the HTTP server for diagnostics, including the health check indicating whether the registration socket exists, will listen (example: `:8080`). The default is empty string, which means the server is disabled. Only one of `--health-port` and `--http-endpoint` can be set.")
 	showVersion             = flag.Bool("version", false, "Show version.")
-	version                 = "unknown"
+
+	// kubelet registration succeed flags
+	kubeletRegistrationAckPath  = flag.String("kubelet-registration-ack-path", "", "If set, a temp file with this name will be created after the kubelet registration process succeeds.")
+	kubeletRegistrationCheckAck = flag.String("kubelet-registration-check-ack", "", "Checks that the kubelet plugin registration ack file exists, if set it must be the same value as kubelet-registration-ack-path.")
```

Comment:
could only use one value `kubeletRegistrationAckPath`, if `kubeletRegistrationAckPath` is set, then always check where kubelet plugin registration succeeds or not, that would be simpler to configure.

Reply:
yes, that was my initial take but I found some problems with the name of the lock file, so I'm checking whether a lock file exists or not (per driver) and to do that this flag would need to have a value so the probe could be:

```
          livenessProbe:
            exec:
              command:
              - /csi-node-driver-registrar
              - "--kubelet-registration-ack-path=my-driver-lock-file"
            initialDelaySeconds: 2
```

The container needs to create this file on startup:

```
      containers:
        - name: csi-driver-registrar
          image: gcr.io/mauriciopoppe-gke-dev/csi-node-driver-registrar:canary
          args:
            - "--kubelet-registration-ack-path=my-driver-lock-file
            - "--kubelet-registration-path=/var/lib/kubelet/plugins/pd.csi.storage.gke.io/csi.sock"
```

The logic to use it would be:

- if it's the only flag set then we'll use it as a probe
- if it's together with `--kubelet-registration-path` then it's not a probe and we should use it to create the file

The fact that there are two behaviors for the same flag felt kinda weird but I could try this approach too

cc @msau42 @jingxu97

Reply:
I've updated the flags usage, now there's one flag to set the lock file path and another one to set the mode:

```
      containers:
        - name: csi-driver-registrar
          image: gcr.io/mauriciopoppe-gke-dev/csi-node-driver-registrar:canary
          args:
            - "--kubelet-registration-succeeded-lock-file=my-driver-lock-file"
            - "--kubelet-registration-path=/var/lib/kubelet/plugins/pd.csi.storage.gke.io/csi.sock"
          livenessProbe:
            exec:
              command:
              - /csi-node-driver-registrar
              - "--kubelet-registration-succeeded-lock-file=my-driver-lock-file"
              - "--kubelet-registration-succeeded-mode=probe"
            initialDelaySeconds: 2
```

The logic to use them is:

- to enable checking the registration process set `--kubelet-registration-succeeded-lock-file` to the path of a file that will be created in an emptydir, this value must be set in both the image args and the liveness probe exec binary args with the same value
- in addition set `--kubelet-registration-succeeded-mode=probe` in the liveness probe

cc @msau42 @jingxu97

Reply:
Do we need the lock file arg? Can we have a predefined file name based on the driver name?

Reply:
the driver name is available a few lines after the fix and it's fetched through an RPC call, I could go through this path once and cache it in memory for later calls of the probe https://github.com/kubernetes-csi/node-driver-registrar/blob/master/cmd/csi-node-driver-registrar/main.go#L148